src/lib/payment-processing/checkoutSession.ts

```diff
@@ -0,0 +1,262 @@
+
+  if (minDate && maxDate && minDate.getDate() === maxDate.getDate()) {
+    shippingName += ` (Estimated delivery: ${formatDate(maxDate)})`;
+  } else if (minDate && maxDate) {
+    shippingName += ` (Estimated delivery: ${formatDate(minDate)}-${formatDate(maxDate)})`;
+  } else if (minDate) {
+    shippingName += ` (Estimated delivery: ${formatDate(minDate)})`;
+  } else if (maxDate) {
+    shippingName += ` (Estimated delivery: ${formatDate(maxDate)})`;
+  }
+
+  return shippingName;
+};
+
+const formatRecipientInfo = (shippingAddress: ShippingAddress): App.Recipient => {
+  // Clone shippingAddress
+  shippingAddress = { ...shippingAddress };
+
+  const callingCodeCountry = shippingAddress.calling_code?.split('+').at(0) as CountryCode;
+  delete shippingAddress.calling_code;
+
+  let recipientInformation = shippingAddress as App.Recipient;
+
+  if (recipientInformation.state_code === 'N/A') {
+    const { state_code, ...rest } = recipientInformation;
+    recipientInformation = rest;
+  }
+
+  if (recipientInformation.phone) {
+    const parsedNumber = parsePhoneNumber(recipientInformation.phone, callingCodeCountry);
+    recipientInformation.phone = parsedNumber.formatInternational();
+  }
+
+  return recipientInformation;
+};
+
+const hydrateItemsFromDb = async (
+  items: CartRequestBody['items']
+): Promise<Array<HydratedCartItem>> => {
+  // Get product variant details from DB
+  const { variants } = await sdk.Variants({
+    variantIds: items.map((v) => v.id)
+  });
+
+  const typedVariants = variants ?? ([] as Array<StrongVariant>);
+
+  // TODO: (maybe) return error if could not find all products in DB
+  // Maybe user had super old session data hanging around after products have been changed... probably not very likely given that this is sessionStorage.
+
+  // Create lookup table using variant ID as the key
+  const variantDetailsHash: Record<string, Omit<StrongVariant, 'id'>> = typedVariants.reduce(
+    (accumulatorHash, { id, ...rest }) => {
+      return {
+        ...accumulatorHash,
+        [id]: rest
+      };
+    },
+    {}
+  );
+
+  return items.map((item) => ({
+    id: item.id,
+    ...variantDetailsHash[item.id],
+    quantity: Number(item.quantity)
+  }));
+};
+
+const getShippingRates = async (
+  items: Array<HydratedCartItem>,
+  recipientInformation: App.Recipient
+): Promise<Array<ShippingRate>> => {
+  const itemsForPrintfulShipping = items.map((item) => {
+    return {
+      variant_id: item.details.baseVariantId,
+      quantity: item.quantity
+    };
+  });
+
+  const shippingRates = await printfulApi('/shipping/rates', {
+    method: 'POST',
+    body: JSON.stringify({
+      recipient: recipientInformation,
+      items: itemsForPrintfulShipping
+    })
+  });
+
+  return shippingRates.map((rate: PrintfulShippingRate) => ({
+    ...rate,
+    displayName: getCustomShippingName(rate)
+  })) as Array<ShippingRate>;
+};
+
```

Comment:
nit - looks like GH doesn't like how I'm indenting this

```suggestion
# The reason we encrypt this data is because we don't want to retain the shipping data longer than
# necessary so the easiest way to do this is to pass it through by the service providers such that once
# payment is complete we can decrypt it and get rid of it and the decryption key while still being able
# to pass it to printful after the payment is completed.
```



Helpful comment to explain what's going on in better detail here since it's a relatively key architectural decision that we made here